File: ios/Packages/HapiKit/Sources/HapiProtocol/Chat/ReducerEvents.swift
import Foundation

// Port of web/src/chat/reducerEvents.ts — CLI pipe-format limit parsing,
// consecutive-event dedupe, api-error folding.

private let limitReachedRegex = JSRegex("\\AClaude AI usage limit reached\\|(\\d+)(?:\\|([^|]*))?\\z")
private let limitWarningRegex = JSRegex("\\AClaude AI usage limit warning\\|(\\d+)\\|(\\d+)\\|([^|]*)\\z")

/// Port of `parseClaudeUsageLimit`.
func parseClaudeUsageLimit(_ text: String) -> AgentEvent? {
    if let match = limitReachedRegex.firstMatch(in: text), let rawTimestamp = match[1] {
        if let timestamp = Double(rawTimestamp), timestamp.isFinite {
            // `reachedMatch[2] || ''` — an absent or empty capture stays ''.
            return .limitReached(endsAt: timestamp, limitType: match.count > 2 ? (match[2] ?? "") : "")
        }
    }

    if let match = limitWarningRegex.firstMatch(in: text),
       let rawTimestamp = match[1], let rawUtilization = match[2] {
        if let timestamp = Double(rawTimestamp), timestamp.isFinite,
           let utilizationInt = Double(rawUtilization), utilizationInt.isFinite {
            let limitType = match.count > 3 ? (match[3] ?? "") : ""
            return .limitWarning(utilization: utilizationInt / 100, endsAt: timestamp, limitType: limitType)
        }
    }

    return nil
}

/// Port of `parseMessageAsEvent`.
func parseMessageAsEvent(_ msg: NormalizedMessage) -> AgentEvent? {
    if msg.isSidechain { return nil }
    guard let content = msg.agentContent else { return nil }

    for item in content {
        if case .text(let text) = item {
            if let limitEvent = parseClaudeUsageLimit(text.text) {
                return limitEvent
            }
        }
    }

    return nil
}

/// Port of `dedupeAgentEvents`: drops a consecutive repeat of the same
/// title-changed/message/error event (and, generically, any byte-identical
/// event), plus a message event echoing the immediately preceding
/// title-changed title.
public func dedupeAgentEvents(_ blocks: [ChatBlock]) -> [ChatBlock] {
    var result: [ChatBlock] = []
    var prevEventKey: String?
    var prevTitleChangedTo: String?

    for block in blocks {
        guard case .agentEvent(let eventBlock) = block else {
            result.append(block)
            prevEventKey = nil
            prevTitleChangedTo = nil
            continue
        }

        let event = eventBlock.event
        if event.type == "title-changed", let title = event["title"]?.stringValue {
            let trimmed = title.jsTrimmed
            let key = "title-changed:\(trimmed)"
            if key == prevEventKey {
                continue
            }
            result.append(block)
            prevEventKey = key
            prevTitleChangedTo = trimmed
            continue
        }

        if event.type == "message", let message = event["message"]?.stringValue {
            let trimmed = message.jsTrimmed
            let key = "message:\(trimmed)"
            if key == prevEventKey {
                continue
            }
            if let prevTitle = prevTitleChangedTo, trimmed == prevTitle {
                continue
            }
            result.append(block)
            prevEventKey = key
            prevTitleChangedTo = nil
            continue
        }

        if event.type == "error", let message = event["message"]?.stringValue {
            let trimmed = message.jsTrimmed
            let key = "error:\(trimmed)"
            if key == prevEventKey {
                continue
            }
            result.append(block)
            prevEventKey = key
            prevTitleChangedTo = nil
            continue
        }

        // Generic key: TS uses JSON.stringify(event); the canonical
        // serialization is its deterministic stand-in (fixture inputs are
        // key-sorted, so the reference's insertion order is sorted too).
        let key = "event:\(toCanonicalJSON(event.wireValue))"

        if key == prevEventKey {
            continue
        }

        result.append(block)
        prevEventKey = key
        prevTitleChangedTo = nil
    }

    return result
}

/// Port of `foldApiErrorEvents`: consecutive api-error events keep only the
/// latest state.
public func foldApiErrorEvents(_ blocks: [ChatBlock]) -> [ChatBlock] {
    var result: [ChatBlock] = []

    for block in blocks {
        guard case .agentEvent(let eventBlock) = block else {
            result.append(block)
            continue
        }

        if eventBlock.event.type != "api-error" {
            result.append(block)
            continue
        }

        if let last = result.last, case .agentEvent(let prev) = last, prev.event.type == "api-error" {
            result[result.count - 1] = block
        } else {
            result.append(block)
        }
    }

    return result
}

File: android/core/protocol/src/main/kotlin/app/hapi/protocol/chat/ReducerEvents.kt
package app.hapi.protocol.chat

import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/** Port of `web/src/chat/reducerEvents.ts`. */

private val LIMIT_REACHED_REGEX = Regex("^Claude AI usage limit reached\\|(\\d+)(?:\\|([^|]*))?$")
private val LIMIT_WARNING_REGEX = Regex("^Claude AI usage limit warning\\|(\\d+)\\|(\\d+)\\|([^|]*)$")

/** JS `Number.parseInt(s, 10)` for `\d+` captures (falls to Double past Long range). */
private fun parsePipeInt(text: String): Double? =
    text.toLongOrNull()?.toDouble() ?: text.toDoubleOrNull()

private fun parseClaudeUsageLimit(text: String): AgentEvent? {
    val reachedMatch = LIMIT_REACHED_REGEX.find(text)
    if (reachedMatch != null) {
        val timestamp = parsePipeInt(reachedMatch.groupValues[1])
        if (timestamp != null && timestamp.isFinite()) {
            return AgentEvent.of(buildJsonObject {
                put("type", "limit-reached")
                put("endsAt", jsNumber(timestamp))
                // TS `reachedMatch[2] || ''` — an unmatched or empty group becomes ''.
                put("limitType", reachedMatch.groupValues[2])
            })
        }
    }

    val warningMatch = LIMIT_WARNING_REGEX.find(text)
    if (warningMatch != null) {
        val timestamp = parsePipeInt(warningMatch.groupValues[1])
        val utilizationInt = parsePipeInt(warningMatch.groupValues[2])
        val limitType = warningMatch.groupValues[3]
        if (timestamp != null && utilizationInt != null) {
            return AgentEvent.of(buildJsonObject {
                put("type", "limit-warning")
                put("utilization", jsNumber(utilizationInt / 100.0))
                put("endsAt", jsNumber(timestamp))
                put("limitType", limitType)
            })
        }
    }

    return null
}

fun parseMessageAsEvent(msg: NormalizedMessage): AgentEvent? {
    if (msg.isSidechain) return null
    val agent = msg as? NormalizedMessage.Agent ?: return null

    for (content in agent.content) {
        if (content is NormalizedAgentContent.Text) {
            val limitEvent = parseClaudeUsageLimit(content.text)
            if (limitEvent != null) {
                return limitEvent
            }
        }
    }

    return null
}

fun dedupeAgentEvents(blocks: List<ChatBlock>): List<ChatBlock> {
    val result = mutableListOf<ChatBlock>()
    var prevEventKey: String? = null
    var prevTitleChangedTo: String? = null

    for (block in blocks) {
        if (block !is AgentEventBlock) {
            result.add(block)
            prevEventKey = null
            prevTitleChangedTo = null
            continue
        }

        val event = block.event
        if (event is AgentEvent.TitleChanged && event.title != null) {
            val title = event.title!!.trim()
            val key = "title-changed:$title"
            if (key == prevEventKey) {
                continue
            }
            result.add(block)
            prevEventKey = key
            prevTitleChangedTo = title
            continue
        }

        if (event is AgentEvent.Message && event.message != null) {
            val message = event.message!!.trim()
            val key = "message:$message"
            if (key == prevEventKey) {
                continue
            }
            if (prevTitleChangedTo != null && message == prevTitleChangedTo) {
                continue
            }
            result.add(block)
            prevEventKey = key
            prevTitleChangedTo = null
            continue
        }

        if (event is AgentEvent.ErrorEvent && event.message != null) {
            val message = event.message!!.trim()
            val key = "error:$message"
            if (key == prevEventKey) {
                continue
            }
            result.add(block)
            prevEventKey = key
            prevTitleChangedTo = null
            continue
        }

        // TS: `event:${JSON.stringify(event)}` — kotlinx JsonObject.toString()
        // is insertion-ordered standard JSON, an equality-faithful analogue.
        val key = "event:${event.raw}"

        if (key == prevEventKey) {
            continue
        }

        result.add(block)
        prevEventKey = key
        prevTitleChangedTo = null
    }

    return result
}

/** Fold consecutive api-error events, keeping only the latest state. */
fun foldApiErrorEvents(blocks: List<ChatBlock>): List<ChatBlock> {
    val result = mutableListOf<ChatBlock>()

    for (block in blocks) {
        if (block !is AgentEventBlock) {
            result.add(block)
            continue
        }

        if (block.event.type != "api-error") {
            result.add(block)
            continue
        }

        val prev = result.lastOrNull()
        if (prev is AgentEventBlock && prev.event.type == "api-error") {
            result[result.size - 1] = block
        } else {
            result.add(block)
        }
    }

    return result
}
